File: client/src/main/java/com/msc/serverbrowser/util/GTAController.java
package com.msc.serverbrowser.util;

import java.awt.Desktop;
import java.awt.Desktop.Action;
import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.Objects;
import java.util.Optional;
import java.util.logging.Level;

import com.github.plushaze.traynotification.animations.Animations;
import com.github.plushaze.traynotification.notification.NotificationTypeImplementations;
import com.github.plushaze.traynotification.notification.TrayNotificationBuilder;
import com.github.sarxos.winreg.HKey;
import com.github.sarxos.winreg.RegistryException;
import com.github.sarxos.winreg.WindowsRegistry;
import com.msc.serverbrowser.data.PastUsernames;
import com.msc.serverbrowser.data.properties.ClientPropertiesController;
import com.msc.serverbrowser.data.properties.Property;
import com.msc.serverbrowser.gui.SAMPVersion;
import com.msc.serverbrowser.logging.Logging;
import com.msc.serverbrowser.util.windows.OSUtility;

import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.util.Duration;

/**
 * Contains utility methods for interacting with native samp stuff.
 *
 * @author Marcel
 */
public final class GTAController
{
	/**
	 * Holds the users username.
	 */
	public static StringProperty usernameProperty = new SimpleStringProperty(retrieveUsernameFromRegistry());

	private GTAController()
	{
		// Constructor to prevent instantiation
	}

	/**
	 * Writes the actual username (from registry) into the past usernames list and sets the new name
	 */
	public static void applyUsername()
	{
		if (!OSUtility.isWindows())
		{
			return;
		}

		killSamp();
		PastUsernames.addPastUsername(retrieveUsernameFromRegistry());
		try
		{
			WindowsRegistry.getInstance().writeStringValue(HKey.HKCU, "SOFTWARE\\SAMP", "PlayerName", usernameProperty.get());
		}
		catch (final RegistryException e)
		{
			Logging.log(Level.WARNING, "Couldn't set username.", e);
		}

	}

	// TODO Think of a better solution
	/**
	 * Returns the Username that samp has set in the registry.
	 *
	 * @return Username or "404 name not found"
	 */
	private static String retrieveUsernameFromRegistry()
	{
		if (!OSUtility.isWindows())
		{
			return "You are on Linux ;D";
		}

		try
		{
			return WindowsRegistry.getInstance().readString(HKey.HKCU, "SOFTWARE\\SAMP", "PlayerName");
		}
		catch (final Exception exception)
		{
			Logging.log(Level.WARNING, "Couldn't retrieve Username from registry.", exception);
			return "404 Name not found";
		}
	}

	/**
	 * Returns the GTA path.
	 *
	 * @return {@link Optional} of GTA path or an empty {@link Optional} if GTA couldn't be found
	 */
	public static Optional<String> getGtaPath()
	{
		if (!OSUtility.isWindows())
		{
			return Optional.of("You are on Linux ;D");
		}

		final String property = ClientPropertiesController.getPropertyAsString(Property.SAMP_PATH);

		if (Objects.isNull(property) || property.isEmpty())
		{
			return Optional.empty();
		}
		return property.endsWith("\\") ? Optional.of(property) : Optional.of(property + "\\");
	}

	/**
	 * Should only be used if necessary.
	 *
	 * @return String of the GTA Path or null.
	 */
	public static String getGtaPathUnsafe()
	{
		if (!OSUtility.isWindows())
		{
			return "You are on Linux ;D";
		}

		try
		{
			return WindowsRegistry.getInstance().readString(HKey.HKCU, "SOFTWARE\\SAMP", "gta_sa_exe").replace("gta_sa.exe", "");
		}
		catch (final RegistryException exception)
		{
			Logging.log(Level.WARNING, "Couldn't retrieve GTA path.", exception);
			return null;
		}
	}

	/**
	 * Returns the version number of the installed samp version, if samp and gta have been found.
	 *
	 * @return {@link Optional} of installed versions version number or an empty {@link Optional}
	 */
	public static Optional<SAMPVersion> getInstalledVersion()
	{
		if (!OSUtility.isWindows())
		{// OS not supported
			return Optional.empty();
		}

		final Optional<String> path = getGtaPath();
		if (!path.isPresent())
		{// GTA couldn't be found
			return Optional.empty();
		}

		final File file = new File(path.get() + "samp.dll");
		if (!file.exists())
		{// samp.dll doesn't exist, even though GTA is installed at this point.
			return Optional.empty();
		}

		/*
		 * Bad Practice, will cause an error if Kalcor decides to do a huge update someday :P
		 */
		switch ((int) file.length())
		{
			case 2199552:
				return Optional.of(SAMPVersion.ZeroThreeSeven);
			case 1093632:
				return Optional.of(SAMPVersion.ZeroThreeZ);
			case 2084864:
				return Optional.of(SAMPVersion.ZeroThreeX);
			case 1998848:
				return Optional.of(SAMPVersion.ZeroThreeE);
			case 2015232:
				return Optional.of(SAMPVersion.ZeroThreeD);
			case 1511424:
				return Optional.of(SAMPVersion.ZeroThreeC);
			case 610304:
				return Optional.of(SAMPVersion.ZeroThreeA);
			default:
				return Optional.empty();
		}

	}

	/**
	 * Connects to the given server (IP and Port) using an empty (no) password. Other than
	 * {@link GTAController#connectToServer(String)} and
	 * {@link GTAController#connectToServer(String, String)}, this method uses the
	 * <code>samp://</code> protocol to connect to make the samp launcher connect to the server.
	 *
	 * @param ipAndPort
	 *            the server to connect to
	 * @return true if it was most likely successful
	 */
	private static boolean connectToServerUsingProtocol(final String ipAndPort)
	{
		if (!OSUtility.isWindows())
		{
			return false;
		}

		try
		{
			Logging.log(Level.INFO, "Connecting using protocol.");
			final Desktop desktop = Desktop.getDesktop();

			if (desktop.isSupported(Action.BROWSE))
			{
				desktop.browse(new URI("samp://" + ipAndPort));
				return true;
			}
		}
		catch (final IOException | URISyntaxException exception)
		{
			Logging.log(Level.WARNING, "Error connecting to server.", exception);
		}

		return false;
	}

	/**
	 * Kills SA-MP using the command line.
	 */
	public static void killSamp()
	{
		kill("samp.exe");
	}

	/**
	 * Kills GTA using the command line.
	 */
	public static void killGTA()
	{
		kill("gta_sa.exe");
	}

	/**
	 * Kills a process with a given name.
	 *
	 * @param processName
	 *            the name that determines what processes will be killed
	 */
	private static void kill(final String processName)
	{
		if (!OSUtility.isWindows())
		{
			return;
		}

		try
		{
			Runtime.getRuntime().exec("taskkill /F /IM " + processName);
		}
		catch (final IOException exception)
		{
			Logging.log(Level.SEVERE, "Couldn't kill " + processName, exception);
		}
	}

	/**
	 * Connects to the given server (IP and Port) using the given password. Uses the commandline to
	 * open samp and connect to the server.
	 *
	 * @param ipAndPort
	 *            the server to connect to
	 * @param password
	 *            the password to use for connecting
	 */
	public static void connectToServer(final String ipAndPort, final String password)
	{
		if (ClientPropertiesController.getPropertyAsBoolean(Property.ALLOW_CLOSE_GTA))
		{
			killGTA();
		}

		final Optional<String> gtaPath = getGtaPath();
		if (gtaPath.isPresent())
		{
			try
			{
				Logging.log(Level.INFO, "Connecting using executeable.");
				final ProcessBuilder builder = new ProcessBuilder(gtaPath.get() + File.separator + "samp.exe ", ipAndPort, password);
				builder.directory(new File(gtaPath.get()));
				builder.start();
			}
			catch (final Exception exception)
			{
				if (Objects.isNull(password) || password.isEmpty())
				{
					connectToServerUsingProtocol(ipAndPort);
				}
				else
				{
					Logging.log(Level.WARNING, "Couldn't connect to server", exception);
				}
			}
		}
		else
		{
			new TrayNotificationBuilder()
					.type(NotificationTypeImplementations.ERROR)
					.title("GTA couldn't be located")
					.message("If this isn't correct, please head to the settings view and manually enter your GTA path.")
					.animation(Animations.POPUP)
					.build().showAndDismiss(Duration.seconds(10));
		}
	}

	/**
	 * Connects to the given server (IP and Port) using an empty (no) password.
	 *
	 * @param ipAndPort
	 *            the server to connect to
	 * @see GTAController#connectToServer(String, String)
	 */
	public static void connectToServer(final String ipAndPort)
	{
		connectToServer(ipAndPort, "");
	}
}

File: src/main/kotlin/com/msc/serverbrowser/util/samp/GTAController.kt
package com.msc.serverbrowser.util.samp

import com.github.plushaze.traynotification.animations.Animations
import com.github.plushaze.traynotification.notification.NotificationTypeImplementations
import com.github.plushaze.traynotification.notification.TrayNotificationBuilder
import com.msc.serverbrowser.Client
import com.msc.serverbrowser.data.PastUsernames
import com.msc.serverbrowser.data.insallationcandidates.InstallationCandidate
import com.msc.serverbrowser.data.properties.ClientPropertiesController
import com.msc.serverbrowser.data.properties.SampPathProperty
import com.msc.serverbrowser.gui.View
import com.msc.serverbrowser.gui.controllers.implementations.VersionChangeController
import com.msc.serverbrowser.severe
import com.msc.serverbrowser.util.basic.HashingUtility
import com.msc.serverbrowser.util.basic.StringUtility
import com.msc.serverbrowser.util.unix.WineUtility
import com.msc.serverbrowser.util.windows.OSUtility
import com.msc.serverbrowser.util.windows.Registry
import com.msc.serverbrowser.warn
import javafx.beans.property.SimpleStringProperty
import javafx.beans.property.StringProperty
import javafx.scene.control.Alert
import javafx.scene.control.Alert.AlertType
import javafx.scene.control.ButtonType
import javafx.scene.control.TextInputDialog
import java.io.File
import java.io.IOException
import java.security.NoSuchAlgorithmException
import java.util.Objects
import java.util.Optional

/**
 * Contains utility methods for interacting with native samp stuff.
 *
 * @author Marcel
 */
object GTAController {
    /**
     * Holds the users username.
     */
    val usernameProperty: StringProperty = SimpleStringProperty(retrieveUsernameFromRegistry() ?: "")

    /**
     * Returns the path to GTA, eventually the path will be resolved if the current OS is not Windows.
     *
     * If a custom path has been set, that will be returned instead.
     */
    val gtaPath: String?
        get() {
            val savedProperty = ClientPropertiesController.getProperty(SampPathProperty)

            if (savedProperty.isNotBlank()) {
                return if (savedProperty.endsWith(File.separator)) {
                    savedProperty
                } else {
                    savedProperty + File.separator
                }
            }

            return gtaPathFromRegistry
        }

    /**
     * Returns the path to GTA without resolving it with `winepath`, no matter which OS the client runs on.
     *
     * If a custom path has been set, that will be returned instead.
     */
    val windowsStyleGtaPath: String?
        get() {
            val savedProperty = ClientPropertiesController.getProperty(SampPathProperty)

            if (savedProperty.isNotBlank()) {
                return if (savedProperty.endsWith(File.separator)) {
                    savedProperty
                } else {
                    savedProperty + File.separator
                }
            }

            return windowsGtaPathFromRegistry
        }

    private val windowsGtaPathFromRegistry: String?
        get() {
            return Registry
                    .readString("HKCU\\Software\\SAMP", "gta_sa_exe")
                    ?.replace("gta_sa.exe", "")
        }

    private val gtaPathFromRegistry: String?
        get() {
            val retrievedPath = windowsGtaPathFromRegistry

            if (OSUtility.isWindows) {
                return retrievedPath
            }

            if (retrievedPath != null) {
                return WineUtility.convertPath(retrievedPath)
            }

            return retrievedPath
        }

    /**
     * Returns the [InstallationCandidate] value that represents the currently installed samp
     * version.
     *
     * @return [Optional] of installed versions version number or an [Optional.empty]
     */
    // GTA couldn't be found samp.dll doesn't exist, even though GTA is installed at this point.
    val installedVersion: Optional<InstallationCandidate>
        get() {
            val path = gtaPath ?: return Optional.empty()

            val file = File(path + "samp.dll")
            if (file.exists().not()) {
                return Optional.empty()
            }

            try {
                val hashSum = HashingUtility.generateChecksum(file.toString())
                return VersionChangeController.INSTALLATION_CANDIDATES.stream()
                        .filter { candidate -> candidate.sampDllChecksum.equals(hashSum, ignoreCase = true) }
                        .findFirst()
            } catch (exception: NoSuchAlgorithmException) {
                severe("Error hashing installed samp.dll", exception)
            } catch (exception: IOException) {
                severe("Error hashing installed samp.dll", exception)
            }

            return Optional.empty()
        }

    /**
     * Writes the actual username (from registry) into the past usernames list and sets the new name
     */
    fun applyUsername() {

        //TODO Implement for linux
        if (OSUtility.isWindows) {
            killSAMP()
        }

        retrieveUsernameFromRegistry()
                .takeIf { it != usernameProperty.get() }
                ?.apply(PastUsernames::addPastUsername)

        Registry.writeString("HKCU\\Software\\SAMP", "PlayerName", usernameProperty.get())
    }

    // TODO Think of a better solution
    /**
     * Returns the Username that samp has set in the registry.
     *
     * @return Username or "404 name not found"
     */
    internal fun retrieveUsernameFromRegistry() = Registry.readString("HKCU\\Software\\SAMP", "PlayerName")

    /**
     * Connects to a server, depending on if it is passworded, the user will be asked to enter a
     * password. If the server is not reachable the user can not connect.
     *
     * @param address server address
     * @param port server port
     * @param serverPassword the password to be used for this connection
     */
    fun tryToConnect(client: Client, address: String, port: Int, serverPassword: String) {
        try {
            SampQuery(address, port).use { query ->
                val serverInfo = query.basicServerInfo

                if (Objects.isNull(serverPassword) || serverPassword.isEmpty() && serverInfo.isPresent && StringUtility.stringToBoolean(serverInfo.get()[0])) {
                    val passwordOptional = promptUserForServerPassword(client)
                    passwordOptional.ifPresent { password -> SAMPLauncher.connect(client, address, port, password) }
                } else {
                    SAMPLauncher.connect(client, address, port, serverPassword)
                }
            }
        } catch (exception: IOException) {
            warn("Couldn't connect to server.", exception)

            if (askUserIfHeWantsToConnectAnyways(client)) {
                SAMPLauncher.connect(client, address, port, serverPassword)
            }
        }

    }

    private fun askUserIfHeWantsToConnectAnyways(client: Client): Boolean {
        val alert = Alert(AlertType.CONFIRMATION, Client.getString("serverMightBeOfflineConnectAnyways"), ButtonType.YES, ButtonType.NO)
        alert.title = Client.getString("connectingToServer")
        alert.initOwner(client.stage!!)
        return alert.showAndWait().orElse(ButtonType.NO) == ButtonType.YES
    }

    /**
     * Shows a dialog prompting the user for a server password.
     *
     * @return an [Optional] containing either a string (empty or filled) or
     * [Optional.empty]
     */
    fun promptUserForServerPassword(client: Client): Optional<String> {
        val dialog = TextInputDialog()
        dialog.initOwner(client.stage!!)
        dialog.title = Client.getString("connectToServer")
        dialog.headerText = Client.getString("enterServerPasswordMessage")

        return dialog.showAndWait()
    }

    /**
     * Shows a TrayNotification that states, that connecting to the server wasn't possible.
     */
    fun showCantConnectToServerError() {
        TrayNotificationBuilder()
                .type(NotificationTypeImplementations.ERROR)
                .title(Client.getString("cantConnect"))
                .message(Client.getString("addressNotValid"))
                .animation(Animations.POPUP)
                .build().showAndDismiss(Client.DEFAULT_TRAY_DISMISS_TIME)
    }

    /**
     * Kills SA-MP using the command line.
     */
    fun killSAMP() {
        kill("samp.exe")
    }

    /**
     * Kills GTA using the command line.
     */
    fun killGTA() {
        kill("gta_sa.exe")
    }

    /**
     * Kills a process with a given name.
     *
     * @param processName the name that determines what processes will be killed
     */
    private fun kill(processName: String) {
        if (!OSUtility.isWindows) {
            return
        }

        try {
            Runtime.getRuntime().exec("taskkill /F /IM $processName")
        } catch (exception: IOException) {
            severe("Couldn't kill $processName", exception)
        }

    }

    /**
     * Displays a notification that states, that GTA couldn't be located and links the Settings page.
     */
    fun displayCantLocateGTANotification(client: Client) {
        val trayNotification = TrayNotificationBuilder()
                .type(NotificationTypeImplementations.ERROR)
                .title(Client.getString("cantFindGTA"))
                .message(Client.getString("locateGTAManually"))
                .animation(Animations.POPUP).build()

        trayNotification.setOnMouseClicked { _ ->
            client.loadView(View.SETTINGS)
            client.settingsController?.selectSampPathTextField()
        }
        trayNotification.showAndDismiss(Client.DEFAULT_TRAY_DISMISS_TIME)
    }
}
